### PR TITLE
Fix r_line_readchar_utf8

### PR DESCRIPTION
**Pull request by MaskRay**
The old code does not handle UTF-8 characters. You will see the prompt being reprinted if you type a character whose codepoint is larger than 0x7f.

It is fixed with this commit:
```
[0x00000000]> ?e 一
```

### CODE REVIEW

File: libr/cons/dietline.c [hunk 1]
```diff
@@ -112,34 +112,32 @@ R_API int r_line_dietline_init() {
...
-		}
-		if (s[len] < 28) {
-			return s[0]? 1: 0;
-		}
-		if (is_valid_char (s[len])) {
-			return s[0]? 1: 0;
-		}
-		if ((s[len] & 0xc0) != 0x80) {
-			continue;
+	ssize_t len, t, i;
```
Comment (radare) [inline libr/cons/dietline.c]: you are not catching ^Z here

Comment (radare) [inline libr/cons/dietline.c]: so my guess is that ^Z doesnt works after your patch

File: libr/cons/dietline.c [hunk 2]
```diff
@@ -112,34 +112,32 @@ R_API int r_line_dietline_init() {
...
+		len = 2;
+	} else if ((s[0] & 0xf0) == 0xe0) {
+		len = 3;
+	} else if ((s[0] & 0xf8) == 0xf0) {
+		len = 4;
+	} else {
+		return -1;
+	}
+	for (i = 1; i < len; i++) {
+		if ((t = read (0, s + i, 1)) < 1) {
```
Comment (radare) [inline libr/cons/dietline.c]: better use != 1

File: libr/cons/dietline.c [hunk 3]
```diff
@@ -112,34 +112,32 @@ R_API int r_line_dietline_init() {
 
 #if USE_UTF8
 /* read utf8 char into 's', return the length in bytes */
-static int r_line_readchar_utf8(unsigned char *s, int slen) {
+static int r_line_readchar_utf8(unsigned char *s) {
```
Comment (radare) [inline libr/cons/dietline.c]: use ut8 instead of unsigned char

Comment (radare) [inline libr/cons/dietline.c]: why are you removing the slen? your code is overflowing

File: libr/cons/dietline.c (hunk 1 above)
Comment (MaskRay) [inline libr/cons/dietline.c]: Done. The way ^Z is handled seems very strange.

File: libr/cons/dietline.c (hunk 2 above)
Comment (MaskRay) [inline libr/cons/dietline.c]: `> 1` is impossible. Anyway, I've changed to `!= 1`

File: libr/cons/dietline.c (hunk 3 above)
Comment (MaskRay) [inline libr/cons/dietline.c]: The call site has 10 bytes for `buf`. Anyway, readded.

**Comment by radare** [comment]
It is :)

**Comment by radare** [comment]
I know but constant comprisons are better than range ones

**Pull request merged**